doc: add qemu tutorial (#1863)
canonical/cloud-init: doc/rtd/topics/tutorials/qemu.rst

deploy it into a Qemu_ virtual machine. We'll be using Qemu for this tutorial
which is a popular emulator for running virtual machines on Linux. Several
popular virtual machine tools use Qemu, including Libvirt, LXD, and Vagrant.

Install Qemu
============

.. code-block:: sh

    $ sudo apt install qemu-system-x86

See qemu's `install instructions <https://www.qemu.org/download/#linux>`_.

Download a Cloud Image
======================

All commands should be executed from a single directory; some
commands create files that are used by later commands. One might want to
use a temporary directory for easy clean up afterwards.

.. code-block:: sh

    $ wget https://cloud-images.ubuntu.com/jammy/current/jammy-server-cloudimg-amd64.img

Cloud images have cloud-init pre-installed.

Define our user data
====================

Create the following file ``user-data``.

.. code-block:: sh

    $ cat << EOF > user-data
    #cloud-config
    password: passw0rd
    chpasswd:
      expire: False

    EOF

Define our meta data
====================

Create the following file on your local filesystem at ``meta-data``.

.. code-block:: sh

    $ cat << EOF > meta-data

>     instance-id: id-007

Author: +1

    local-hostname: jammy

    EOF

Define our vendor data
======================

Not necessary, but faster (retry wait time).

.. code-block:: sh

    $ touch vendor-data

Start an ad hoc IMDS Server
===========================

.. code-block:: sh

    $ python3 -m http.server --directory . &

Launch a vm with our user data
==============================

Now that we have LXD setup and our user data defined, we can launch an
instance with our user data:

.. code-block:: sh

    $ qemu-system-x86_64                                            \
        -net nic                                                    \
        -net user                                                   \

>         -machine accel=kvm,type=q35                                 \

Author: > do you strictly need q35 type?

No

> almost everywhere "as fast as it can"

As a tutorial I think compatibility with various realease/distros is more important than speed. Fallback to tcg makes sense here. +1 thanks

.. _tutorial_qemu:

> Qemu

Reviewer: We may need to make the title a bit more descriptive. Where it appears in the LHS menu as just "Qemu", a new user may not know what it is or whether it's relevant to their interests.

Author: +1

****

> In this tutorial, we will create our first cloud-init user data config and

Reviewer: I'd rework this paragraph a bit to first briefly explain what Qemu is and why someone might want to use it for deploying cloud-init, and then telling the user what end goal the tutorial will walk them towards.

Author: +1

deploy it into a Qemu_ virtual machine. We'll be using Qemu for this tutorial
which is a popular emulator for running virtual machines on Linux. Several
popular virtual machine tools use Qemu, including Libvirt, LXD, and Vagrant.

Install Qemu
============

.. code-block:: sh

>     $ sudo apt install qemu-system-x86

Reviewer: Is it conventional to use the $ signs in the code snippets? A couple of times they caused me problems when I was copy/pasting them.

Author: Yes, this is a convention. It helps to disambiguate whether a command is single-line or multi-line, and is sometimes used to indicate non-root vs root commands (using $ and #, respectively).

Consider the heredoc commands below (line 41). The $ here helps users to understand that these are not individual commands to be executed in sequence, but are part of a multi-line shell command.

I've seen this question come up enough times in IRC/Stack Exchange/etc that I agree it is potentially worth dropping to avoid confusion. If not for the heredoc commands (see line 41), I think this would be easy to eliminate them all.

My concern is that if I saw the following while skimming a *nix tutorial, I would have a moment of pause because lines after the first one are not valid shell commands. 
```
    cat << EOF > user-data
    #cloud-config
    password: passw0rd
    chpasswd:
      expire: False
    EOF
```
Is this more readable to others? If we're alright with this format I'd be fine with dropping all of the $s.

Thoughts?

Reviewer 2: It does also indicate what lines are command and what are output:

```
$ pwd
/home/aciba
```

And prevents from disaster while copying and pasting commands, if I copy-paste `$ rm /` and press enter, then I will get an error because of the invalid `$`.

> See qemu's `install instructions <https://www.qemu.org/download/#linux>`_.

Reviewer: ```suggestion
If you are not using Ubuntu, you can visit Qemu's `install instructions <https://www.qemu.org/download/#linux>`_ for additional information.
```

> Download a Cloud Image

Author: I don't believe so.

======================

> All commands should be executed from a single directory; some

Reviewer: If it's generally considered a good idea to create a temporary directory, then for a tutorial doc this should be presented as an explicit step rather than an optional one. It wasn't completely clear to me if I had to download the image into the temp directory, so I didn't create one. It didn't occur to me until after that maybe I should have done that.

Author: +1

commands create files that are used by later commands. One might want to
use a temporary directory for easy clean up afterwards.

.. code-block:: sh

    $ wget https://cloud-images.ubuntu.com/jammy/current/jammy-server-cloudimg-amd64.img

Cloud images have cloud-init pre-installed.

Define our user data
====================

> Create the following file ``user-data``.

Reviewer: ```suggestion
Create a file called ``user-data`` in your temporary directory and copy/paste the following code into it.
```

Author: I've made this confusing, sorry.

The text here describes what the code does, not what the reader is supposed to do. If one creates a file named `user-data` and then copies in the following contents, the tutorial will not work.

The tutorial is written such that all the user should have to do is copy, paste, and execute each code block.

I'll add some clarifying instructions at the top to make this clear.

Reviewer: This is probably how I managed to break the tutorial, to be fair...lol

.. code-block:: sh

    $ cat << EOF > user-data
    #cloud-config
    password: password
    chpasswd:
      expire: False

    EOF

Define our meta data
====================

> Create the following file on your local filesystem at ``meta-data``.

Author: Same comment as [this](https://github.com/canonical/cloud-init/pull/1863/files#r1034030434).

.. code-block:: sh

    $ cat << EOF > meta-data
    instance-id: id-007
    local-hostname: jammy

    EOF

Define our vendor data
======================

> Not necessary, but faster as it avoids retry wait time looking for it.

Reviewer: ```suggestion
Now create the empty file ``vendor-data`` in your temporary directory. This will speed up the retry wait time.
```

Author: +1 I like the clarity this adds.

.. code-block:: sh

    $ touch vendor-data

> Start an ad hoc IMDS Server

Reviewer: I have no idea what an IMDS Server is or why I'd want one - this section needs a bit of context if it's aimed at a new user.

Author: Fair enough :D

===========================

.. code-block:: sh

    $ python3 -m http.server --directory . &

> Launch a vm with our user data

Reviewer: ```suggestion
Launch a VM with our user data
```

> Launch a vm with our user data

Reviewer: ```suggestion
Launch a virtual machine (VM) with our user data
```

==============================

> Now that we have LXD setup and our user data defined, we can launch an

Author: Good catch! I'll fix that. LXD shouldn't be mentioned in this tutorial.

instance with our user data:

.. code-block:: sh

    $ qemu-system-x86_64                                            \
        -net nic                                                    \
        -net user                                                   \
        -machine accel=kvm,type=q35                                 \
        -cpu host                                                   \
        -m 512                                                      \
        -nographic                                                  \
        -hda jammy-server-cloudimg-amd64.img                        \
        -smbios type=1,serial=ds='nocloud-net;s=http://10.0.2.2:8000/'

Verify that cloud-init ran successfully
=======================================

> After launching the virtual machine we should be able to connect to our

Reviewer: It wasn't clear to me that I had to press enter after the VM was launched to make the login bit appear

Author: Yeah that's confusing? I forgot about that.

I think what's happening there is that when the prompt is ready it prints to stdout, but since systemd and kernel logs are still printing the old prompt gets lost in the mix (if you scroll up you might see it). Stdin is listening, even if you don't see the prompt, but hitting enter will make it print again.

I'll add something to make it clear to users.

instance using the default distro username. In this case the default username
is ``ubuntu`` and the password we configured is ``password``

> If you can log in using the configured password, it worked!

Author: I think that [this comment](https://github.com/canonical/cloud-init/pull/1863/files#r1034030434) might explain what happened, however you raise a good point that some help for how to debug this would be helpful.

> If you can log in using the configured password, it worked!

Reviewer: It may not be necessary to add debug info, if you're already making the step I messed up clearer (if we can avoid someone doing the stupid thing I did, they wouldn't need to debug in the tutorial)

Check the cloud-init status:

.. code-block:: sh

    $ cloud-init status --wait
    .....
    cloud-init status: done

Tear down
=========

> Exit the qemu shell using ``ctrl-a x`` (that's ctrl and a

Reviewer: ```suggestion
Exit the qemu shell using ``ctrl-a x`` (that's ``ctrl`` and ``a``
```

simultaneously, followed by ``x``).

If you started the python webserver in the background (using ``&``),

> then don't forget to bring it to the foreground (``fg``) and kill it

Reviewer: I think we're supposed to avoid using words like "kill" - is there an alternative we can use here (cancel, quit, exit, etc).

Author: +1

(``ctrl-c``).

What's next?
============

> In this tutorial, we configured the default user's password.

Reviewer: Did we do that with a module?

Author: I'll add a comment.

The full list of modules available can be found in
:ref:`modules documentation<modules>`.

> Each module contains examples of how to use it.

Reviewer: ```suggestion
The documentation for each module contains examples of how to use it.
```

.. _tutorial_qemu:

Qemu Tutorial
*************

.. toctree::
   :titlesonly:
   :hidden:

   qemu-debugging.rst

> Tutorial overview
> =================

Reviewer: ```suggestion
```

Reviewer: This looks a bit odd when rendered in RTD because it's so close to the page title. I'd recommend just taking it out, since it's not really needed.

In this tutorial, we will demonstrate launching an Ubuntu cloud image in a
virtual machine that uses cloud-init to pre-configure the system during boot.

The goal of this tutorial is to provide a minimal demonstration of cloud-init
that users can use as a development environment to test cloud-init

> configurations locally prior to launching in the cloud.

Reviewer: Nice - this is exactly what I was looking for :D just be careful not to switch from talking to the user (as in the first sentence) to talking about them (in this one).

Why Qemu?
=========

> Qemu is a cross-platform emulator capable of running performant virtual

Reviewer: This is really good - I would just put a link to the Qemu website here at this first proper mention of it (for convenience)

Author: +1

machines. Qemu is used at the core of a broad range of production operating
system deployments and open source software projects (including libvirt, LXD,
and vagrant) and is capable of running Windows, Linux, and Unix guest operating
systems. While Qemu is flexibile and feature-rich, we are using it because of
the broad support it has due to its broad adoption and ability to run on
\*nix-derived operating systems.

What is an IMDS?
================

Instance Metadata Service is a service provided by most cloud providers
as a means of providing information to virtual machine instances. This

> service is used by cloud providers to surface information to a virtual machine.

Reviewer: Not sure if this is my lack of knowledge here - I'm a bit confused by the use of "surface" as a verb. Is this a common terminology that we can assume the users will know/understand?

Author: **I** use this term a lot, though this is probably more colloquial than it needs to be. I'll use "to expose" instead.

This service is used for many different things, and is the primary
mechanism for some clouds to surface cloud-init configuration data to
the instance.

How does cloud-init use the IMDS?
=================================

> The IMDS contains a private http webserver to each operating

Reviewer: ```suggestion
The IMDS contains a private http webserver to each operating
```

> The IMDS contains a private http webserver to each operating

Reviewer: "contains to" feels a bit odd to my eyes. Again I'm not sure if this is my lack of knowledge at play.

Author: +1

system instance launched. During early boot, cloud-init sets up network
access and queries this webserver to gather configuration data. This allows
cloud-init to configure your operating system while it boots.

In this tutorial we emulate this workflow using Qemu and a simple python
webserver. This workflow may be suitable for developing and testing cloud-init
configurations prior to cloud deployments.

How to use this tutorial
========================

In this tutorial each code block is to be copied and pasted directly
into the terminal then executed. Omit the prompt ``$`` before each command.

Each code block is preceded by a description of what the command does.

Install Qemu
============

Install Qemu.

.. code-block:: sh

    $ sudo apt install qemu-system-x86

If you are not using Ubuntu, you can visit Qemu's `install instructions`_ for
additional information.

Create a temporary directory
============================

This directory will store our cloud image and configuration files,
``meta-data`` [add-link], ``vendor-data`` [add-link], and ``user-data``

> [add-link],

Author: Oops, forgot this note to self (will add a minor patch for this tomorrow).

Cloud images typically come with cloud-init pre-installed and configured to run
on first boot. Users should not need to worry about installing cloud-init
unless they are manually creating their own images. In this case we select the
latest Ubuntu LTS_.

Download the server image using ``wget``.

.. code-block:: sh

    $ wget https://cloud-images.ubuntu.com/jammy/current/jammy-server-cloudimg-amd64.img

Define our user data
====================

Create the following file ``user-data``. This user-data cloud-config
sets the password of the default user and sets it to never expire. For
more details see this module_.

Execute the following command, which creates a file named ``user-data`` with
configuration data.

.. code-block:: sh

    $ cat << EOF > user-data
    #cloud-config
    password: password
    chpasswd:
      expire: False

    EOF

What is user data?
==================

Before moving forward, let's inspect our user data file.

.. code-block:: sh

   $ cat user-data

You should see the following contents:

.. code-block:: yaml

    #cloud-config
    password: password
    chpasswd:
      expire: False

The first line starts with ``#cloud-config``, which tells cloud-init

> what kind of configuration is contained. The cloud-config config type uses yaml

Reviewer: ```suggestion
what kind of configuration is contained. The cloud-config config type use yaml
```

Author: I'm going to push back a little on this change request. Since this config "type" is singular wouldn't `use` make more sense here?

I'm happy to make changes for clarity, but I do think that the current grammar makes more sense (which leads me to think that something about this sentence was confusing).

Thoughts?

Reviewer: You're quite right, grammatically your original wording is correct, I have no idea why I thought it wasn't - I think the double "config" might have made my brain short circuit for a hot minute. If there is an alternative way of phrasing that sentence without changing your meaning, it might be a good idea to change it. It's a bit confusing to look at.

Author: > If there is an alternative way of phrasing that sentence without changing your meaning, it might be a good idea to change it. It's a bit confusing to look at.

How is this wording? https://github.com/canonical/cloud-init/pull/1863/commits/8398443f79ca46ce25beee74cb7f0eba23046f49

Reviewer: Much better! Clear, but still concise - good job +1

Multiple different formats are supported by cloud-init. See the
:ref:`documentation describing different formats<user_data_formats>`.

The second line, ``password: password``, per :ref:`the docs<mod-users_groups>`,
sets the default user's password to ``password``.

The third and fourth lines direct cloud-init to set this default password to
never expire.

Define our meta data
====================

Execute the following command, which creates a file named ``meta-data`` with
configuration data.

.. code-block:: sh

    $ cat << EOF > meta-data
    instance-id: someid/somehostname
    local-hostname: jammy

    EOF

Define our vendor data
======================

Now create the empty file ``vendor-data`` in your temporary directory. This
will speed up the retry wait time.

.. code-block:: sh

    $ touch vendor-data

Start an ad hoc IMDS Server
===========================

In a separate terminal, change to your temporary directory and then start the
python webserver (built-in to python).

.. code-block:: sh

    $ cd temp
    $ python3 -m http.server --directory .

Launch a virtual machine with our user data
===========================================

> Launch the virtual machine. By default, qemu will print to the terminal both
> kernel logs and systemd logs while the operating system boots. This may take a
> few moments to complete.

Reviewer: ```suggestion
Launch the virtual machine. By default, Qemu will print the kernel logs
and systemd logs to the terminal while the operating system boots. This
may take a few moments to complete.
```

Reviewer: Just to make it completely crystal clear, I'd preface this little bit by saying something like "In our original terminal window" or something like that. Idiot that I am, I sat wondering which I was meant to use for a good minute until I realised that I couldn't use the IMDS window while the server was running.

> If the output stopped scrolling but you don't see a prompt yet, type ``enter``

Reviewer: ```suggestion
If the output stopped scrolling but you don't see a prompt yet, press ``enter``
```

Reviewer: I got a bit confused by the code block appearing below this. I'd swap the order (so the thing about the prompt appears below the code block).

to get to login prompt.

.. code-block:: sh

    $ qemu-system-x86_64                                            \
        -net nic                                                    \
        -net user                                                   \
        -machine accel=kvm:tcg                                      \
        -cpu host                                                   \
        -m 512                                                      \
        -nographic                                                  \
        -hda jammy-server-cloudimg-amd64.img                        \
        -smbios type=1,serial=ds='nocloud-net;s=http://10.0.2.2:8000/'

Verify that cloud-init ran successfully
=======================================

After launching the virtual machine we should be able to connect to our
instance using the default distro username.

In this case the default username is ``ubuntu`` and the password we configured
is ``password``.

If you can log in using the configured password, it worked!

If you cloudn't log in, see

> :ref:`this page for debug information<debug information>`.

Reviewer: ```suggestion
:ref:`this page for debug information<qemu_debug_info>`.
```
